ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -493,119 +493,117 @@ let vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far t
-                         mr_local_xenops_locator = vconf.xenops_locator;
-                         mr_remote_xenops_locator = Xapi_xenops.xenops_vdi_locator_of_strings dest_sr_uuid remote_vdi;
-                         mr_local_vdi_reference = vconf.vdi;
-                         mr_remote_vdi_reference = remote_vdi_reference; }) in
-  try
-    let result = continuation mirror_record in
-    
-    (match mirror_id with
-     | Some mid -> ignore(Storage_access.unregister_mirror mid);
-     | None -> ());
-              
-    if mirror && not (Xapi_fist.storage_motion_keep_vdi () || copy)
-    then
-      Helpers.call_api_functions ~__context (fun rpc session_id ->
+  let post_mirror ?mirror_id mirror_record =
+    try
+      let result = continuation mirror_record in
+      (match mirror_id with
+      | Some mid -> ignore(Storage_access.unregister_mirror mid);
+      | None -> ());
+      if mirror && not (Xapi_fist.storage_motion_keep_vdi () || copy) then
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
           XenAPI.VDI.destroy rpc session_id vconf.vdi);
-    
-    result
-  with e ->
-    (* Stop mirroring and check to see if it was us that caused the mirror failure *)
-    let mirror_failed = 
-      match mirror_id with
-      | Some mid ->
-        (try SMAPI.DATA.MIRROR.stop ~dbg ~id:mid with _ -> ());
-        let m = SMAPI.DATA.MIRROR.stat ~dbg ~id:mid in
-        m.Mirror.failed
-      | None ->
-        false
-    in
-
-    (* Now that mirroring has finished, clean up any datapath we made *)
-    (match newdp with | Some dp -> (try SMAPI.DP.destroy ~dbg ~dp ~allow_leak:false with _ -> error "Failed to cleanup datapath: %s" dp) | None -> ());
-
-    (* And destroy the new VDI *)
-    (try XenAPI.VDI.destroy remote.rpc remote.session remote_vdi_reference with _ -> error "Failed to destroy remote VDI");
-
-    if mirror_failed then raise (Api_errors.Server_error(Api_errors.mirror_failed,[Ref.string_of vconf.vdi]));
-    
-    raise e
+      result
+    with e ->
+      let mirror_failed =
+        match mirror_id with
+        | Some mid ->
+           ignore(Storage_access.unregister_mirror mid);
+          (try SMAPI.DATA.MIRROR.stop ~dbg ~id:mid with _ -> ());
+          let m = SMAPI.DATA.MIRROR.stat ~dbg ~id:mid in
+          m.Mirror.failed
+        | None -> false in
+      if mirror_failed then raise (Api_errors.Server_error(Api_errors.mirror_failed,[Ref.string_of vconf.vdi]))
+      else raise e in
+
+  if mirror then
+    with_new_dp (fun new_dp ->
+      let mirror_id, remote_vdi = mirror_to_remote new_dp in
+      with_remote_vdi remote_vdi (fun remote_vdi_ref ->
+        let mirror_record = get_mirror_record ~new_dp remote_vdi remote_vdi_ref in
+        post_mirror ?mirror_id mirror_record))
+  else
+    let mirror_record = get_mirror_record vconf.location (XenAPI.VDI.get_by_uuid remote.rpc remote.session vdi_uuid) in
+    post_mirror mirror_record
```

Comment:
Yes, with the current code, the effect is the same.

If we want to put more common logic in post_mirror in the future, making a uniform call might make better sense. For now, indeed, the semantic is no difference.

Let me know if you'd prefer calling the continuation directly here, in which case I'll make a quick update.